docs/cs-proof.md

```diff
@@ -0,0 +1,800 @@
+Constraint System Proofs
+========================
+
+_Constraint system_ is a system that defines arithmetic constraints over a set of variables.
+The proof system allows _a verifier_ to specify the constraints, for which _a prover_ is asked to generate a valid proof.
+The resulting proof is _zero-knowledge_: while the constraints are known to both the prover and the verifier, the variables remain secret.
+
+The constraint system we use is specifically a _rank-1 quadratic constraint system_ or _R1CS_.
+Such system allows expressing linear constraints over _individual variables_ as well as _multiplications of variables_.
+
+The proving system uses efficient [inner product protocol](../inner_product_proof/index.html) by expressing all the constraints in terms of a single inner product.
+The following notes describe in detail how this is accomplished.
+
+Notation
+--------
+
+Dimensions of vectors:
+
+* \\(m\\) — number of secret values \\({\mathbf{v}}\\),
+* \\(n\\) — number of variable multiplication gates represented by \\(\mathbf{a}\_L, \mathbf{a}\_R, \mathbf{a}\_O\\),
+* \\(q\\) — number of linear constraints represented by \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O, \mathbf{W}\_V\\).
+
+In the [Bulletproofs paper][bp_website], matrices are labeled as \\(\mathbf{W}\_L, \mathbf{W}\_R, \mathbf{W}\_O, \mathbf{W}\_V\\). We will keep this notation, but will note to readers not to confuse the \\(\mathbf{W}\_{L,R,O,V}\\) notation for being a vector of points.
+
+We will use the notation described in the [`notation`](../notes/index.html#notation) section of the notes. In addition, we rename two more variables from the paper for better consistency with the rest of the notation:
+\\[
+\begin{aligned}
+    \beta         &\xrightarrow{} \tilde{o} \\\\
+	Q             &\xrightarrow{} q \\\\
+\end{aligned}
+\\]
+
+[bp_website]: https://crypto.stanford.edu/bulletproofs/
+
+Constraint system
+-----------------
+
+The constraint system consists of two sets of constraints: [multiplication gates](#multiplication-gates) and [linear constraints](#linear-constraints) in terms of the [variables](#variables).
+
+### Variables
+
+There are two kinds of variables in the constraint system:
+
+1. \\(m\\) **high-level** variables, the input secrets \\(\mathbf{v}\\),
+2. \\(n\\) **low-level** variables, the internal inputs and outputs of the [multiplication gates](#multiplication-gates).
```

Comment:
maybe mention here that these correspond to a_L, a_R, a_O, for clarity (so you can know that before reading ahead to the multiplication gates)

Reply:
👍